src/java.base/share/classes/java/util/UUID.java

```diff
@@ -99,9 +107,178 @@ public final class UUID implements java.io.Serializable, Comparable<UUID> {
     /*
      * The random number generator used by this class to create random
      * based UUIDs. In a holder class to defer initialization until needed.
+     * <p>
+     * The implementation buffers the reads to avoid bashing SecureRandom with
+     * small requests. It also maintains a SecureRandom per buffer to alleviate
+     * scalability bottlenecks when reading from a single (synchronized) SecureRandom.
      */
-    private static class Holder {
-        static final SecureRandom numberGenerator = new SecureRandom();
+    private static final class RandomUUID {
+        static final String PROP_NAME_PRNG_NAME = "java.util.UUID.prngName";
+        static final String PROP_NAME_BUF_COUNT = "java.util.UUID.buffersCount";
+
+        static final String PRNG_NAME;
+        static final int BUFS_COUNT;
+        static final Buffer[] BUFS;
+
+        private static final int roundPowerOfTwo(int x) {
+            int n = -1 >>> Integer.numberOfLeadingZeros(x - 1);
+            return (n < 0) ? 1 : (n + 1);
+        }
+
+        static {
+            try {
+                PRNG_NAME = GetPropertyAction.privilegedGetProperty(PROP_NAME_PRNG_NAME, null);
+                try {
+                    newRandom();
+                } catch (Exception e) {
+                    throw new IllegalArgumentException(PROP_NAME_PRNG_NAME + " is incorrect", e);
+                }
+                int bufCount = GetIntegerAction.privilegedGetProperty(PROP_NAME_BUF_COUNT,
+                                    Runtime.getRuntime().availableProcessors());
+                if (bufCount < 1) {
+                    throw new IllegalArgumentException(PROP_NAME_BUF_COUNT + " is out of range");
+                }
+                BUFS_COUNT = roundPowerOfTwo(bufCount);
+                BUFS = new Buffer[BUFS_COUNT];
+            } catch (Exception e) {
+                throw new InternalError(e);
```

Comment:
Would it be better to throw `ExceptionInInitializerError` here?

Reply:
Sure, see new commit.